test/sumlog.jl

```diff
@@ -0,0 +1,7 @@
+@testset "sumlog" begin
+    for T in [Int, Float16, Float32, Float64, BigFloat]
```

Comment:
I noticed that you removed the type restriction. Thus we should extend the tests and eg. check more general iterables (also with different types, abstract eltype etc since `sum(log, x)` would work for them) and also complex numbers.

Reply:
`eltype` doesn't work well for `Base.Generator`s. Usually this is when I'd turn to something like
```julia
julia> Core.Compiler.return_type(gen.f, Tuple{eltype(gen.iter)})
Float64
```
We could instead have it fall back on the default, but I'd guess that will sacrifice performance.

Reply:
I bet you could write an equally fast version which explicitly calls `iterate`, and widens if the type changes. (But usually the compiler will prove that it won't.)

One reason to keep `mapreduce` for arrays is that you can give it `dims`.

Reply:
Should check more carefully, but this appears to work & is as fast as current version:
```julia
function sumlog(x)
    iter = iterate(x)
    if isnothing(iter)
        return eltype(x) <: Number ? zero(float(eltype(x))) : 0.0
    end
    x1 = float(iter[1])
    x1 isa AbstractFloat || return sum(log, x)
    sig, ex = significand(x1), exponent(x1)
    iter = iterate(x, iter[2])
    while iter !== nothing
        xj = float(iter[1])
        x1 isa AbstractFloat || return sum(log, x)  # maybe not ideal, re-starts iterator
        sig, ex = _sumlog_op((sig, ex), (significand(xj), exponent(xj)))
        iter = iterate(x, iter[2])
    end
    return log(sig) + IrrationalConstants.logtwo * ex
end

sumlog(f, x) = sumlog(Iterators.map(f, x))
sumlog(f, x, ys...) = sumlog(f(xy...) for xy in zip(x, ys...))
```

And for dims:
```julia
sumlog(x::AbstractArray{T}; dims=:) where T = _sumlog(float(T), dims, x)

function _sumlog(::Type{T}, ::Colon, x) where {T<:AbstractFloat}
    sig, ex = mapreduce(_sumlog_op, x; init=(one(T), zero(exponent(one(T))))) do xj
        float_xj = float(xj)
        significand(float_xj), exponent(float_xj) 
    end
    return log(sig) + IrrationalConstants.logtwo * ex
end

function _sumlog(::Type{T}, dims, x) where {T<:AbstractFloat}
    sig_ex = mapreduce(_sumlog_op, x; dims=dims, init=(one(T), zero(exponent(one(T))))) do xj
        float_xj = float(xj)
        significand(float_xj), exponent(float_xj) 
    end
    map(sig_ex) do (sig, ex)
        log(sig) + IrrationalConstants.logtwo * ex
    end
end
```
Should I make a PR to the PR?

Reply:
https://github.com/cscherrer/LogExpFunctions.jl/pull/1 is a tidier version of the above.